src/librustc_mir/interpret/eval_context.rs

```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+struct SpanGuard(log::span::Span);
```

Comment:
I don't _entirely_ get the motivation behind this type versus just using the `tracing` `Span` type directly...why is this needed? Are there potential upstream changes we can make that would make this type unnecessary?

Reply:
I want do be able to automatically exit the span on `Drop`. When I use `.enter()` I get a guard, but this guard borrows the `Span`, so I can't store both of them in the same datastructure. I could probably manually invoke `enter` and `exit` at the push/pop of the stack frames, but that seems fragile compared to using a guard like this.

Reply:
> I want do be able to automatically exit the span on `Drop`. When I use `.enter()` I get a guard, but this guard borrows the `Span`, so I can't store both of them in the same datastructure.

Ah, okay, that makes sense. A change we could make upstream to better support this use case is adding a method to `Span` that's like `enter`, but takes the span by value, and returns an owned guard. The owned guard would have an explicit `exit` method that returns the unentered span, as well as exiting on drop. I discussed this proposal here: https://github.com/tokio-rs/tracing/issues/79#issuecomment-503276300

This would be quite simple to implement upstream. Thus far, we haven't done so just because there doesn't seem to be a whole lot of demand for it, but this seems like a good use-case.

Reply:
Another note: this type should probably not implement `Send`. Entering a span applies to the _current_ thread; when a span is entered, the thread on which `enter` is called is considered to be in that span. While the `Span` type can be freely sent or shared across threads, the entry should not — if this type is sent to a different thread and then dropped, it will call `exit` on _that_ thread, which may have never entered the span; meanwhile, the thread where the span _was_ entered will remain in that span forever.